### ISSUE TITLE
reduce external dependencies?

### ISSUE THREAD

**Issue by bodsch** [xyproto/algernon#17]
it is possible to reduce the external dependcy like this?
https://github.com/xyproto/algernon/blob/93349dfc4d3daa5351c51551e8a8b9f0633bdc90/themes/html.go#L88

I like your project, but hard-coded external dependencies are not acceptable to me. :}

**Comment by xyproto** [comment]
Thanks for filing the issue! I agree and will look into fixing this.

**Comment by xyproto** [comment]
I'm planning to use [chroma](https://github.com/alecthomas/chroma) serverside instead of highlight.js client side.

**Comment by bodsch** [comment]
i will plan a new docker container for testing in an environment without network access.

**Comment by xyproto** [comment]
The (experimental) [splash](https://github.com/xyproto/splash) package is a new package I created for applying syntax highlighting with `chroma` to existing HTML data. It is used by the `noextdep2` branch.

The `noextdep2` branch should solve the issue with external dependencies, since it should be completelty self-contained.

It has some style issues that highlight.js didn't have. Highlighted lines of code when displaying Markdown files are wrapped, and the manual code style specifications are ignored.

Nevertheless, it should be ready for testing in an environment without network access, while the stylistic issues are being worked out.

**Comment by xyproto** [comment]
* Merged the changes for removing all external dependencies into master.
* Removed the `noextdep2` branch.

**Comment by bodsch** [comment]
my working container: https://github.com/bodsch/docker-algernon
i remove the external references to fonts.googleapis too: https://github.com/bodsch/docker-algernon/blob/master/Dockerfile#L40-L51
thats neccessary for (my) project.

**Comment by xyproto** [comment]
Thanks for creating a docker image for Algernon!

Updated the code to not use external fonts, I somehow missed those.

I also updated all external dependencies to use the new `go mod` command in Go 1.11. The latest dependencies of everything also made Algernon noticeably faster.

**Comment by bodsch** [comment]
great! <3
i will test the conatiner against the master branch.

**Comment by bodsch** [comment]
currently, all my builds failed with vaious errors. :(

**Comment by xyproto** [comment]
I can add back the vendored dependencies.

Make sure to build either outside of GOPATH or with:

    env GO111MODULE=on go build

Like in https://github.com/xyproto/algernon/blob/master/.travis.yml

**Comment by bodsch** [comment]
(you ar sooo fast! ;) )
thanks, i will try this

**Comment by xyproto** [comment]
Added back the `vendor/` directory as well.

**Comment by bodsch** [comment]
my error are this:
```
go: downloading github.com/mattn/go-colorable v0.0.9
build github.com/xyproto/algernon: cannot find module for path gopkg.in/gcfg.v1
```

**Comment by xyproto** [comment]
Will look into it. Which version of go are you using?

**Comment by bodsch** [comment]
```go version go1.11 linux/amd64```

**Comment by xyproto** [comment]
`go mod verify` revealed an issue with `gopkg.in/gcfg.v1`.

Updated the code, please try again.

Travis CI status: https://travis-ci.org/xyproto/algernon/builds/427165077

**Comment by bodsch** [comment]
My current build step looks like this:
```
RUN \
  go version && \
  cd ${GOPATH}/src/github.com/xyproto/algernon && \
  export name=algernon && \
  export version=$(grep -i version main.go | head -1 | cut -d' ' -f4 | cut -d'"' -f1) && \
  GO111MODULE=on go build && \
  # GOOS=linux go build -o $name.linux && \
  GO111MODULE=on go test && \
  mkdir /tmp/algernon && \
  cp -v $name.linux /tmp/algernon/$name
```
And here comes the error.

```
go: downloading github.com/mattn/go-isatty v0.0.4
go: downloading github.com/mattn/go-colorable v0.0.9
# github.com/xyproto/algernon/engine
engine/basic.go:83:36: undefined: blackfriday.MarkdownCommon
engine/rendering.go:46:11: undefined: blackfriday.MarkdownCommon
engine/rendering.go:207:14: undefined: blackfriday.MarkdownCommon
```

**Comment by xyproto** [comment]
The error message is because blackfriday is not at the latest version (as specified in the go.mod and go.sum files, and in the vendor directory). There is probably a different checkout of the blackfriday files somewhere. Perhaps the docker container needs to be built from scratch, discarding any earlier steps?

**Comment by xyproto** [comment]
I have not tested it yet, but I believe `docker build --no-cache` might solve the problem.

**Comment by bodsch** [comment]
I build my container every time from scratch. :(
But, i use some GO environments variables
```
  GOMAXPROCS=4
  GOOS=linux
  GOARCH=amd64
```
the result of my last (step by step) build
```
/opt/go/src/github.com/xyproto/algernon # GO111MODULE=on go mod verify
all modules verified
/opt/go/src/github.com/xyproto/algernon # GO111MODULE=on go build -v
go: downloading gopkg.in/warnings.v0 v0.1.2
Fetching https://gopkg.in/warnings.v0?go-get=1
go: downloading gopkg.in/gcfg.v1 v1.2.3
Fetching https://gopkg.in/gcfg.v1?go-get=1
Parsing meta tags from https://gopkg.in/gcfg.v1?go-get=1 (status code 502)
Parsing meta tags from https://gopkg.in/warnings.v0?go-get=1 (status code 502)
/opt/go/pkg/mod/github.com/go-gcfg/gcfg@v1.2.3/read.go:11:2: unknown import path "gopkg.in/gcfg.v1/scanner": unrecognized import path "gopkg.in/gcfg.v1" (parse https://gopkg.in/gcfg.v1?go-get=1: no go-import meta tags ())
/opt/go/pkg/mod/github.com/go-gcfg/gcfg@v1.2.3/read.go:12:2: unknown import path "gopkg.in/gcfg.v1/token": unrecognized import path "gopkg.in/gcfg.v1" (parse https://gopkg.in/gcfg.v1?go-get=1: no go-import meta tags ())
/opt/go/pkg/mod/github.com/go-gcfg/gcfg@v1.2.3/set.go:14:2: unknown import path "gopkg.in/gcfg.v1/types": unrecognized import path "gopkg.in/gcfg.v1" (parse https://gopkg.in/gcfg.v1?go-get=1: no go-import meta tags ())
/opt/go/pkg/mod/github.com/go-gcfg/gcfg@v1.2.3/errors.go:3:8: unknown import path "gopkg.in/warnings.v0": unrecognized import path "gopkg.in/warnings.v0" (parse https://gopkg.in/warnings.v0?go-get=1: no go-import meta tags ())
/opt/go/src/github.com/xyproto/algernon #
```

I make a break and will resume tomorrow. :)
Thanks for your help!

**Comment by xyproto** [comment]
Yes please, a new issue for this would be great. I'm planning to ask people more knowledgable in the inner workings of the `go mod` command what might be causing this. Thanks for the patience in trying to figure this one out.

**Comment by xyproto** [comment]
When changing `BUILD_TYPE` to `BUILD_TYPE ?= master` in the `Makefile` in [docker-algernon](https://github.com/bodsch/docker-algernon), Algernon builds and the troubles with blackfriday goes away.

I'm planning to make a new release of Algernon too, which should hopefully also make `docker-algernon` build with `BUILD_TYPE ?= stable`.

**Comment by xyproto** [comment]
Released version `1.11.0`, which works here with `BUILD_TYPE ?= stable`.
